### PR TITLE
synchronize access to wal storage from chain and node

### PR DESCRIPTION
**Pull request by Param-S**
#### Type of change
- Bug fix

#### Description
sychronized the access to raft/wal stroage from
node ready loop and chain apply go routine

#### Related issues
#3903 

Signed-off-by: Parameswaran Selvam <parselva@in.ibm.com>

### CODE REVIEW

File: orderer/consensus/etcdraft/chain.go [hunk 1]
```diff
@@ -1149,12 +1149,10 @@ func (c *Chain) apply(ents []raftpb.Entry) {
 				continue
 			}
 
-			if err := c.Node.storage.Sync(); err != nil {
-				c.logger.Debugf("Failed to sync raft log, error: %s", err)
-			}
+			// persist the WAL entries into disk
+			c.Node.storage.WALSyncC <- struct{}{}
```
Comment (yacovm) [inline orderer/consensus/etcdraft/chain.go:1153]: You need to wait for the sync, no?

File: orderer/consensus/etcdraft/storage.go [hunk 2]
```diff
@@ -441,6 +443,7 @@ func (rs *RaftStorage) ApplySnapshot(snap raftpb.Snapshot) {
 
 // Close closes storage
 func (rs *RaftStorage) Close() error {
+	close(rs.WALSyncC)
```
Comment (yacovm) [inline orderer/consensus/etcdraft/storage.go]: why do you close this channel? you are writing to it in line 1153 in `chain.go` isn't that a problem?

File: orderer/consensus/etcdraft/node.go [hunk 3]
```diff
@@ -171,6 +171,11 @@ func (n *node) run(campaign bool) {
 			// to the followers and them writing to their disks. Check 10.2.1 in thesis
 			n.send(rd.Messages)
 
+		case <-n.storage.WALSyncC:
+			if err := n.storage.Sync(); err != nil {
+				n.logger.Debugf("Failed to sync raft log, error: %s", err)
```
Comment (yacovm) [inline orderer/consensus/etcdraft/node.go]: why debug? we would want to know about it if we failed to sync

File: orderer/consensus/etcdraft/chain.go (hunk 1 above)
Comment (Param-S) [reply orderer/consensus/etcdraft/chain.go:1153]: It need not to wait.  This is the reasoning for my statement, now the access to storage all moved to `node.go ready goroutine` and it is sequenced/synchronzied on select, hence, the `chain.go apply goroutine` need not to wait.

File: orderer/consensus/etcdraft/node.go (hunk 3 above)
Comment (Param-S) [reply orderer/consensus/etcdraft/node.go]: Yes, We can make it warn level.

File: orderer/consensus/etcdraft/storage.go (hunk 2 above)
Comment (Param-S) [reply orderer/consensus/etcdraft/storage.go]: The channel is getting closed to release the resource when the chain is halted.

File: orderer/consensus/etcdraft/chain.go (hunk 1 above)
Comment (yacovm) [reply orderer/consensus/etcdraft/chain.go:1153]: But if we don't wait for the sync, then the goroutine of the node can be too late at syncing and in the meanwhile the goroutine of the chain will continue without the sync happening, no? That's not equivalent to the previous code where the sync is synchronous and blocking.

File: orderer/consensus/etcdraft/storage.go (hunk 2 above)
Comment (yacovm) [reply orderer/consensus/etcdraft/storage.go]: To release what resource?

If the chain is still referenced by a goroutine then closing it won't do any good. If it's not referenced by any goroutine then the channel will also be garbage collected. 

But I am asking - what if the chain goroutine will write to the closed channel after it's closed?

File: orderer/consensus/etcdraft/chain.go (hunk 1 above)
Comment (Param-S) [reply orderer/consensus/etcdraft/chain.go:1153]: as per my understanding, the sender is blocked until the receiver receives the value on the channel...
`Receivers always block until there is data to receive. If the channel is unbuffered, the sender blocks until the receiver has received the value.` [ref](https://go.dev/doc/effective_go#channels)

If this is the case, the node goroutine will get its chance to proceed further in concurrent with chain goroutine.
 
Yes, there is chance if the node goroutine is context switched between the channel receive and actual sync operation. Then. It is possible that the chain goroutine continues its operation. 

But, still chain goroutine execution does not conflict with the execution of the node goroutine even if the current config change evicts/halts the chain.    Please correct me if I am missing anything here.

File: orderer/consensus/etcdraft/storage.go (hunk 2 above)
Comment (Param-S) [reply orderer/consensus/etcdraft/storage.go]: Agree. I will remove it.

File: orderer/consensus/etcdraft/chain.go (hunk 1 above)
Comment (yacovm) [reply orderer/consensus/etcdraft/chain.go:1153]: Fine, leave this as it is.

File: orderer/consensus/etcdraft/storage.go (hunk 2 above)
Comment (yacovm) [reply orderer/consensus/etcdraft/storage.go]: And seems like the unit test failed on the exact send:

```

==================
panic: send on closed channel

goroutine 2985 [running]:
github.com/hyperledger/fabric/orderer/consensus/etcdraft.(*Chain).apply(0xc007304300, {0xc0072b9830, 0x1, 0x0?})
	/home/runner/work/fabric/fabric/orderer/consensus/etcdraft/chain.go:1153 +0x52b
```

**Review by yacovm — APPROVED** [review]

**Comment by yacovm** [comment]
@Mergifyio backport release-2.5

**Pull request merged**